File: backend/app/services/voice_processor.py
"""
Voice DSP first-pass processor — jitter, shimmer, and SNR from audio samples.

Approach:
  1. Input: amplitude envelope samples (normalised -1.0 to 1.0) at 4410 Hz.
  2. F0 estimation via zero-crossing rate (proxy for voiced segment period).
  3. Jitter: cycle-to-cycle period variation = std(period) / mean(period) × 100.
  4. Shimmer: cycle-to-cycle amplitude variation = std(amp) / mean(amp) × 100.
  5. SNR proxy: ratio of RMS voiced energy to RMS silence energy.

Input contract:
  audio_samples: list of float in [-1.0, 1.0], sample rate = 4410 Hz.
  A 5-second recording → 22,050 samples (downsampled from 44100 Hz by client).

Output:
  VoiceResult (jitter_pct, shimmer_pct, snr_db, voiced_fraction, flags)

Privacy note:
  Only amplitude samples are sent — no identifiable voice characteristics.
  Processing is fully server-side and deterministic.
"""

from __future__ import annotations

import math
from collections.abc import Sequence
from dataclasses import dataclass, field

import numpy as np
from scipy import signal

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

SAMPLE_RATE = 4410  # Hz (downsampled from 44100)
MIN_SAMPLES = SAMPLE_RATE * 2  # at least 2 seconds
SILENCE_THRESHOLD = 0.01  # amplitude below this is silence
F0_LOW_HZ = 80.0  # minimum F0 — lowest typical voice fundamental
F0_HIGH_HZ = 400.0  # maximum F0 — highest typical voice fundamental
MIN_VOICED_FRACTION = 0.5  # >50% of recording must be voiced
NOISE_FLOOR_DB = -60.0  # dB floor for SNR calculation


# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------


@dataclass
class VoiceResult:
    """Voice DSP-derived wellness indicator values."""

    jitter_pct: float | None  # cycle-to-cycle period variation (%)
    shimmer_pct: float | None  # cycle-to-cycle amplitude variation (%)
    snr_db: float | None  # signal-to-noise ratio in dB
    voiced_fraction: float  # fraction of recording that is voiced
    flags: list[str] = field(default_factory=list)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def process_audio(samples: Sequence[float]) -> VoiceResult:
    """
    Process normalised audio amplitude samples into voice wellness indicators.

    Parameters
    ----------
    samples:
        Sequence of float in [-1.0, 1.0] at SAMPLE_RATE Hz.

    Returns
    -------
    VoiceResult with jitter_pct, shimmer_pct, snr_db.
    """
    flags: list[str] = []
    arr = np.array(samples, dtype=np.float64)

    if len(arr) < MIN_SAMPLES:
        flags.append("insufficient_samples")
        return VoiceResult(
            jitter_pct=None,
            shimmer_pct=None,
            snr_db=None,
            voiced_fraction=0.0,
            flags=flags,
        )

    # -----------------------------------------------------------------------
    # 1. Detect voiced vs silence frames (20ms frames)
    # -----------------------------------------------------------------------
    frame_size = int(SAMPLE_RATE * 0.02)  # 20ms frames
    num_frames = len(arr) // frame_size
    frame_rms = np.array(
        [
            float(np.sqrt(np.mean(arr[i * frame_size : (i + 1) * frame_size] ** 2)))
            for i in range(num_frames)
        ]
    )

    voiced_mask = frame_rms > SILENCE_THRESHOLD
    voiced_fraction = float(np.mean(voiced_mask))

    if voiced_fraction < MIN_VOICED_FRACTION:
        flags.append("insufficient_voiced_content")

    # -----------------------------------------------------------------------
    # 2. SNR proxy: RMS voiced / RMS silence
    # -----------------------------------------------------------------------
    voiced_rms_vals = frame_rms[voiced_mask]
    silence_rms_vals = frame_rms[~voiced_mask]

    snr_db: float | None = None
    if len(voiced_rms_vals) > 0:
        voiced_rms = float(np.mean(voiced_rms_vals))
        if len(silence_rms_vals) > 0 and float(np.mean(silence_rms_vals)) > 1e-8:
            noise_rms = float(np.mean(silence_rms_vals))
            snr_db = round(20.0 * math.log10(voiced_rms / noise_rms), 1)
        else:
            snr_db = 40.0  # No detectable silence → very high SNR

    if snr_db is not None and snr_db < 10.0:
        flags.append("high_noise")

    # -----------------------------------------------------------------------
    # 3. Extract voiced segment for F0 analysis
    # -----------------------------------------------------------------------
    if voiced_fraction < MIN_VOICED_FRACTION or len(voiced_rms_vals) < 5:
        return VoiceResult(
            jitter_pct=None,
            shimmer_pct=None,
            snr_db=snr_db,
            voiced_fraction=voiced_fraction,
            flags=flags,
        )

    # Work on voiced frames only
    voiced_frames_idx = np.where(voiced_mask)[0]
    voiced_samples = np.concatenate(
        [arr[i * frame_size : (i + 1) * frame_size] for i in voiced_frames_idx]
    )

    # -----------------------------------------------------------------------
    # 4. F0 estimation via autocorrelation (more robust than ZCR)
    # -----------------------------------------------------------------------
    # Bandpass filter voiced signal to F0 range
    nyq = SAMPLE_RATE / 2.0
    low = F0_LOW_HZ / nyq
    high = min(F0_HIGH_HZ / nyq, 0.99)

    try:
        b, a = signal.butter(4, [low, high], btype="bandpass")
        voiced_filtered = signal.filtfilt(b, a, voiced_samples)
    except Exception:  # noqa: BLE001
        flags.append("filter_error")
        return VoiceResult(
            jitter_pct=None,
            shimmer_pct=None,
            snr_db=snr_db,
            voiced_fraction=voiced_fraction,
            flags=flags,
        )

    # Find peaks in filtered voiced signal (each peak ≈ one glottal cycle)
    min_period_samples = int(SAMPLE_RATE / F0_HIGH_HZ)  # shortest cycle
    max_period_samples = int(SAMPLE_RATE / F0_LOW_HZ)   # longest cycle

    peaks, properties = signal.find_peaks(
        voiced_filtered,
        distance=min_period_samples,
        prominence=0.005,
    )

    if len(peaks) < 4:
        flags.append("insufficient_glottal_cycles")
        return VoiceResult(
            jitter_pct=None,
            shimmer_pct=None,
            snr_db=snr_db,
            voiced_fraction=voiced_fraction,
            flags=flags,
        )

    # -----------------------------------------------------------------------
    # 5. Jitter = cycle-to-cycle period variation
    # -----------------------------------------------------------------------
    periods = np.diff(peaks).astype(float)  # in samples
    # Filter out unreasonable periods
    valid = (periods >= min_period_samples) & (periods <= max_period_samples)
    periods = periods[valid]

    jitter_pct: float | None = None
    if len(periods) >= 3:
        mean_period = float(np.mean(periods))
        if mean_period > 0:
            jitter_pct = round(float(np.std(periods)) / mean_period * 100.0, 3)

    # -----------------------------------------------------------------------
    # 6. Shimmer = cycle-to-cycle amplitude variation
    # -----------------------------------------------------------------------
    # Amplitude at each detected peak
    peak_amplitudes = np.abs(voiced_filtered[peaks])

    shimmer_pct: float | None = None
    if len(peak_amplitudes) >= 3:
        mean_amp = float(np.mean(peak_amplitudes))
        if mean_amp > 1e-8:
            shimmer_pct = round(float(np.std(peak_amplitudes)) / mean_amp * 100.0, 3)

    return VoiceResult(
        jitter_pct=jitter_pct,
        shimmer_pct=shimmer_pct,
        snr_db=snr_db,
        voiced_fraction=round(voiced_fraction, 3),
        flags=flags,
    )


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def build_audio_samples(raw: list[float]) -> list[float]:
    """
    Validate and clamp audio sample values from API payload.
    Samples must be in [-1.0, 1.0] range.
    """
    return [max(-1.0, min(1.0, float(v))) for v in raw]

File: mobile/src/utils/voiceProcessor.ts
/**
 * On-device voice DSP — jitter, shimmer, and SNR estimation.
 *
 * All processing runs on the device. Only the derived jitter_pct and
 * shimmer_pct scalars are submitted to the backend — audio samples never
 * leave the device.
 *
 * Mirrors the logic from backend/app/services/voice_processor.py.
 */

import { filtfiltHP, filtfiltLP, findPeaks } from './rppgProcessor';

// ─── Constants (match backend voice_processor.py) ────────────────────────────

const SAMPLE_RATE = 4410;
const FRAME_SIZE = 88; // 20 ms at 4410 Hz
const SILENCE_THRESHOLD = 0.01;
const MIN_VOICED_FRACTION = 0.5;
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const F0_LOW_HZ = 80;
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const F0_HIGH_HZ = 400;
const MIN_PERIOD_SAMPLES = Math.floor(4410 / 400); // 11
const MAX_PERIOD_SAMPLES = Math.floor(4410 / 80); // 55
const K_HP_VOICE = Math.exp((-2 * Math.PI * 80) / 4410); // ≈ 0.8923
const ALPHA_LP_VOICE = 1 - Math.exp((-2 * Math.PI * 400) / 4410); // ≈ 0.4341

// ─── Public interface ─────────────────────────────────────────────────────────

export interface VoiceOnDeviceResult {
  jitter_pct: number | null;
  shimmer_pct: number | null;
  snr_db: number | null;
  voiced_fraction: number;
  flags: string[];
}

// ─── Private helpers ──────────────────────────────────────────────────────────

function mean(arr: number[]): number {
  if (arr.length === 0) return 0;
  let s = 0;
  for (let i = 0; i < arr.length; i++) s += arr[i];
  return s / arr.length;
}

function stdDev(arr: number[]): number {
  if (arr.length < 2) return 0;
  const m = mean(arr);
  let s = 0;
  for (let i = 0; i < arr.length; i++) s += (arr[i] - m) ** 2;
  return Math.sqrt(s / arr.length);
}

function round3(v: number): number {
  return Math.round(v * 1000) / 1000;
}

// ─── Main export ──────────────────────────────────────────────────────────────

export function processVoice(samples: number[]): VoiceOnDeviceResult {
  const flags: string[] = [];

  // 1. Guard: need at least 2 seconds of audio
  if (samples.length < SAMPLE_RATE * 2) {
    flags.push('insufficient_samples');
    return { jitter_pct: null, shimmer_pct: null, snr_db: null, voiced_fraction: 0, flags };
  }

  // 2. Frame RMS
  const numFrames = Math.floor(samples.length / FRAME_SIZE);
  const frameRms: number[] = new Array(numFrames);
  for (let f = 0; f < numFrames; f++) {
    const start = f * FRAME_SIZE;
    let sumSq = 0;
    for (let i = start; i < start + FRAME_SIZE; i++) sumSq += samples[i] * samples[i];
    frameRms[f] = Math.sqrt(sumSq / FRAME_SIZE);
  }

  // 3. Voiced mask
  const voicedMask: boolean[] = frameRms.map(rms => rms > SILENCE_THRESHOLD);

  // 4. voiced_fraction
  const voicedCount = voicedMask.filter(Boolean).length;
  const voiced_fraction = numFrames > 0 ? voicedCount / numFrames : 0;

  // 5. SNR
  const voicedRmsList = frameRms.filter((_, i) => voicedMask[i]);
  const silenceRmsList = frameRms.filter((_, i) => !voicedMask[i]);

  let snr_db: number | null;
  if (voicedRmsList.length === 0) {
    snr_db = null;
  } else if (silenceRmsList.length === 0) {
    snr_db = 40; // all voiced — no silence reference
  } else {
    const voicedRmsMean = mean(voicedRmsList);
    const noiseRmsMean = mean(silenceRmsList);
    snr_db = noiseRmsMean > 0 ? 20 * Math.log10(voicedRmsMean / noiseRmsMean) : 40;
  }

  if (snr_db !== null && snr_db < 10) {
    flags.push('high_noise');
  }

  // 6. Insufficient voiced content guard
  if (voiced_fraction < MIN_VOICED_FRACTION) {
    flags.push('insufficient_voiced_content');
    return {
      jitter_pct: null,
      shimmer_pct: null,
      snr_db,
      voiced_fraction: round3(voiced_fraction),
      flags,
    };
  }

  // 7. Concatenate voiced frames
  const voicedSamples: number[] = [];
  for (let f = 0; f < numFrames; f++) {
    if (voicedMask[f]) {
      const start = f * FRAME_SIZE;
      for (let i = start; i < start + FRAME_SIZE; i++) {
        voicedSamples.push(samples[i]);
      }
    }
  }

  // 8. Bandpass filter
  const filtered = filtfiltLP(filtfiltHP(voicedSamples, K_HP_VOICE), ALPHA_LP_VOICE);

  // 9. Peak detection
  const peaks = findPeaks(filtered, MIN_PERIOD_SAMPLES, 0.005);

  // 10. Insufficient glottal cycles guard
  if (peaks.length < 4) {
    flags.push('insufficient_glottal_cycles');
    return {
      jitter_pct: null,
      shimmer_pct: null,
      snr_db,
      voiced_fraction: round3(voiced_fraction),
      flags,
    };
  }

  // 11. Jitter: inter-peak periods
  const periods: number[] = [];
  for (let i = 1; i < peaks.length; i++) {
    const period = peaks[i] - peaks[i - 1];
    if (period >= MIN_PERIOD_SAMPLES && period <= MAX_PERIOD_SAMPLES) {
      periods.push(period);
    }
  }

  let jitter_pct: number | null = null;
  if (periods.length >= 3) {
    const meanPeriod = mean(periods);
    jitter_pct = meanPeriod > 0 ? round3((stdDev(periods) / meanPeriod) * 100) : null;
  }

  // 12. Shimmer: peak amplitudes
  const peakAmps = peaks.map(i => Math.abs(filtered[i]));
  const meanAmp = mean(peakAmps);
  let shimmer_pct: number | null = null;
  if (meanAmp >= 1e-8) {
    shimmer_pct = round3((stdDev(peakAmps) / meanAmp) * 100);
  }

  // 13. Return result
  return {
    jitter_pct,
    shimmer_pct,
    snr_db,
    voiced_fraction: round3(voiced_fraction),
    flags,
  };
}
